connect/api/src/main/java/org/apache/kafka/connect/data/Values.java

```diff
@@ -0,0 +1,1117 @@
+        }
+    }
+
+    protected static void appendIterable(StringBuilder sb, Iterator<?> iter) {
+        if (iter.hasNext()) {
+            append(sb, iter.next(), true);
+            while (iter.hasNext()) {
+                sb.append(',');
+                append(sb, iter.next(), true);
+            }
+        }
+    }
+
+    protected static String escape(String value) {
+        return value.replaceAll("\\\\", "\\\\\\\\").replaceAll("\"", "\\\\\"");
+    }
+
+    protected static DateFormat dateFormatFor(java.util.Date value) {
+        if (value.getTime() < MILLIS_PER_DAY) {
+            return new SimpleDateFormat(ISO_8601_TIME_FORMAT_PATTERN);
+        }
+        if (value.getTime() % MILLIS_PER_DAY == 0) {
+            return new SimpleDateFormat(ISO_8601_DATE_FORMAT_PATTERN);
+        }
+        return new SimpleDateFormat(ISO_8601_TIMESTAMP_FORMAT_PATTERN);
+    }
+
+    protected static SchemaAndValue parse(Parser parser, boolean embedded) throws NoSuchElementException {
+        if (!parser.hasNext()) {
+            return null;
+        }
+        if (embedded) {
+            if (parser.canConsume(NULL_VALUE)) {
+                return null;
+            }
+            if (parser.canConsume(QUOTE_DELIMITER)) {
+                StringBuilder sb = new StringBuilder();
+                while (parser.hasNext()) {
+                    if (parser.canConsume(QUOTE_DELIMITER)) {
+                        break;
+                    }
+                    sb.append(parser.next());
+                }
+                return new SchemaAndValue(Schema.STRING_SCHEMA, sb.toString());
+            }
+        }
+        if (parser.canConsume(TRUE_LITERAL)) {
+            return TRUE_SCHEMA_AND_VALUE;
+        }
+        if (parser.canConsume(FALSE_LITERAL)) {
+            return FALSE_SCHEMA_AND_VALUE;
+        }
+        int startPosition = parser.mark();
+        try {
+            if (parser.canConsume(ARRAY_BEGIN_DELIMITER)) {
+                List<Object> result = new ArrayList<>();
+                Schema elementSchema = null;
+                while (parser.hasNext()) {
+                    if (parser.canConsume(ARRAY_END_DELIMITER)) {
+                        Schema listSchema = elementSchema == null ? null : SchemaBuilder.array(elementSchema).schema();
+                        result = alignListEntriesWithSchema(listSchema, result);
+                        return new SchemaAndValue(listSchema, result);
+                    }
+                    SchemaAndValue element = parse(parser, true);
+                    elementSchema = commonSchemaFor(elementSchema, element);
+                    result.add(element.value());
+                    parser.canConsume(COMMA_DELIMITER);
+                }
+                // Missing either a comma or an end delimiter
+                if (COMMA_DELIMITER.equals(parser.previous())) {
+                    throw new DataException("Malformed array: missing element after ','");
```

Comment:
Could you try running this test:
```
SchemaAndValue arr = Values.parseString("[1, 2, 3,,,]");
// expect an Exception.
```
I think it returns incorrect values. The trailing commas return the value of `parser.original()` from [line 836](https://github.com/apache/kafka/pull/4319/files?diff=unified#diff-fbab5614eb6f1b3a75939c8fb7df6dc2R866). It should return something more local here.